add micropython RP2040 i2c driver

oclyke opened the pull request:
**what it do**
adds an i2c driver that is compatible with the RP2040 port of micropython

**depended on by**
[add support for micropython on RP2040 boards](https://github.com/sparkfun/Qwiic_Py/pull/14)

**status**
✅ i2c driver seems to work

Commit:
Create micropython_rp2040_i2c.py

Commit:
add micropython_rp2040_i2c driver

Commit:
use board module to set i2c pins

also fix a few bugs with usage of bytes/bytarray

Commit:
fix address retransmission

Commit:
remove i2cbus.write() calls

it doesn't like these. just pack data into the one writeto() call

kirk-sfe reviewed (commented):
**__init__.py**
- See comment regarding this file in the qwiic_py pull request. 
- Basically consolidate everything into this file (only one file) and use exception handling to determine what drivers are available for platform evaluation

Use a pattern like this at the start of the file
_drivers=[]
try:
     from .linux_i2c import LinuxI2C
except e:
    pass
else:
    _drivers.append(LinuxI2C)

**MicroPython**
Does this driver need to be specific to the raspberry pi - can't it be generic micropython?

**Naming**
Instead of the rp2040 name, use "pico". This is "platform" name of the RPi microcontrollers - the rp2040 is just initial microcontroller version

sfe-SparkFro committed:
Change RP2040 driver to be generic MicroPython

We can support more than just the RP2040!

sfe-SparkFro committed:
Import platform drivers more carefully

Can't assume every platform driver will be present, surround imports in try blocks

sfe-SparkFro commented:
Updated to be a more generic MicroPython platform driver, not specific to RP2040.

Also updated platform driver imports like suggested to handle cases when other platform drivers are not included.

Will be reviewing this in more detail before merging.

sfe-SparkFro committed:
Simplify MicroPython read/write functions

Use MicroPython's readfrom_mem() and writeto_mem() functions

sfe-SparkFro committed:
Improve I2C driver importing

Store module and class names of all supported platforms in lists
Loop through each platform and attempt to import each, and append to driver list

sfe-SparkFro committed:
Remove "RP2040" from MicroPython header comment

gigapod commented on qwiic_i2c/__init__.py:
```diff
@@ -63,11 +63,24 @@
+# append them to the driver list if so
+for i in range(len(sub_modules)):
+	try:
+		full_name = "qwiic_i2c." + sub_modules[i]
```
You could skip the creation of the full_name variable and just put the RHS of this statement in the call to import.

gigapod commented on qwiic_i2c/__init__.py:
```diff
@@ -63,11 +63,24 @@
+# All supported platform module and class names
+sub_modules = ["linux_i2c", "circuitpy_i2c", "micropython_i2c"]
+class_names = ["LinuxI2C", "CircuitPythonI2C", "MicroPythonI2C"]
+
```
One thing that you could do here - use a dict - where you have the sub-module the key and the contents be the class name .Something like:
```pythong
canidate_drivers = { "linux_i2c" : "LinuxI2C", 
                                   "circuitpy_i2c" : "CircuitPythonI2C",
                                  "micropython_i2c" : "MicroPythonI2C"}
  for mod, obj  in canidate_drivers.items():
      try:
          sub_module = __import__("qwiic_i2c."+mod)
          _drivers.append(getattr(sub_module, obj)
     except:
           pass

```
Or something like that -

sfe-SparkFro committed:
Change platform loop to use a dictionary
